### PR TITLE
Remove AP_InertialNav from AC_Fence

### PR DESCRIPTION
**Pull request by peterbarker**
Also factors out functions for each different fence type

I would prefer this to take an AP_AHRS rather than a NavEKF, but tridge bounced the idea of having get_location on AP_AHRS for the time being.

A relative position is required for running - only need an absolute position for loading.

### CODE REVIEW

**Comment by khancyr** [comment]
LGTM

File: libraries/AC_Fence/AC_Fence.cpp [hunk 1]
```diff
@@ -276,11 +336,10 @@ bool AC_Fence::check_destination_within_fence(const Location_Class& loc)
     // polygon fence check
     if ((get_enabled_fences() & AC_FENCE_TYPE_POLYGON) && _boundary_num_points > 0) {
         // check ekf has a good location
-        Location temp_loc;
-        if (_inav.get_location(temp_loc)) {
-            const struct Location &ekf_origin = _inav.get_origin();
-            Vector2f position = location_diff(ekf_origin, loc) * 100.0f;
-            if (_poly_loader.boundary_breached(position, _boundary_num_points, _boundary, true)) {
+        Vector2f posNE;
+        if (_ahrs.get_relative_position_NE_origin(posNE)) {
```
Comment (OXINARF) [inline libraries/AC_Fence/AC_Fence.cpp]: Copied from #7345:

> The previous code was a bit strange so I see why you got this wrong. See that the `get_location` call is with a `temp_loc` variable that is never used later. You need to get the origin (and check it is valid) and then do the `location_diff` and use that Vector for checking the boundary.

**Comment by OXINARF** [comment]
> A relative position is required for running - only need an absolute position for loading.

@peterbarker And how is it guaranteed that loading has happened? I don't see that anywhere but maybe I missed it.

File: libraries/AC_Fence/AC_Fence.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AC_Fence/AC_Fence.cpp]: Confused.

On load we translate the boundary to origin-relative coordinates.

On check we check using our origin-relative position.

This does work in testing - which it probably couldn't if this was being screwed up?

File: libraries/AC_Fence/AC_Fence.cpp [hunk 2]
```diff
@@ -145,6 +145,54 @@ bool AC_Fence::pre_arm_check(const char* &fail_msg) const
...
+        // not enabled; no breach
+        return false;
+    }
+
+    // check consistency of number of points
+    if (_boundary_num_points != _total) {
+        // we have no idea where the fence is.  Can't breach it?!
+        _boundary_loaded = false;
+        load_polygon_from_eeprom();
+        return false;
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/AC_Fence/AC_Fence.cpp]: Still seems reasonable to fail prearms if the fence isn't valid?

Also, which check were you expecting to trigger?  Looks like a boundary can be loaded but invalid - and the within-check doesn't look to be doing sanity checks...

File: libraries/AC_Fence/AC_Fence.cpp [hunk 3]
```diff
@@ -130,15 +130,15 @@ bool AC_Fence::pre_arm_check(const char* &fail_msg) const
...
-    if (!_ahrs.get_filter_status(filt_status)) {
-        fail_msg = "AHRS status not available";
-        return false;
-    }
-    if ((_enabled_fences & (AC_FENCE_TYPE_CIRCLE | AC_FENCE_TYPE_POLYGON))>0 && !filt_status.flags.horiz_pos_abs && !filt_status.flags.pred_horiz_pos_abs) {
-        fail_msg = "fence requires position";
-        return false;
+    // if we have horizontal limits enabled, check we can get a
+    // relative position from the EKF
+    if ((_enabled_fences & AC_FENCE_TYPE_CIRCLE) ||
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/AC_Fence/AC_Fence.cpp]: OK, changing this to `_home`

Pretty sure we should be able to split some of these checks in the future.

**Comment by peterbarker** [comment]
> @peterbarker And how is it guaranteed that loading has happened? I don't see that anywhere but maybe I missed it.

`_boundary_valid`?

**Comment by peterbarker** [comment]
@OXINARF @khancyr I've pushed up patches based on @OXINARF 's review.

I might also have slipped in some SYS_STATUS bits...

File: libraries/AC_Fence/AC_Fence.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/AC_Fence/AC_Fence.cpp]: > Pretty sure we should be able to split some of these checks in the future.

I have nothing against that.

File: libraries/AC_Fence/AC_Fence.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/AC_Fence/AC_Fence.cpp]: > Still seems reasonable to fail prearms if the fence isn't valid?

This isn't in the pre-arm checks...

> Also, which check were you expecting to trigger? Looks like a boundary can be loaded but invalid - and the within-check doesn't look to be doing sanity checks...

Right after this `return false;` there is a `if (!_boundary_valid) return false;`. which is what says if the boundary is valid or not.

File: libraries/AC_Fence/AC_Fence.cpp (hunk 1 above)
Comment (OXINARF) [reply libraries/AC_Fence/AC_Fence.cpp]: Please look at the code again. This code is in a method with the following signature `bool check_destination_within_fence(const Location_Class& loc)`. So where is this code checking that `loc` is inside the fence? What this new code is checking is that the current position is inside the fence - that's not what this method is for!

**Comment by OXINARF** [comment]
I'm completely surprised to see this merged. The bugs I identified are still present and were merged into master. I understand that people want to do work on top of this, but merging code with identified bugs just makes my job useless.

> _boundary_valid?

@peterbarker Sure, where is that in the pre-arm checks?

**Comment by rmackay9** [comment]
I wasn't aware there were any bugs not addressed.. hopefully @peterbarker can address them..

**Comment by peterbarker** [comment]
@OXINARF To be clear - the `_boundary_valid` thing is a pre-existing bug, right?

File: libraries/AC_Fence/AC_Fence.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AC_Fence/AC_Fence.cpp]: OK, now I understand.  PR coming.

**Comment by OXINARF** [comment]
@peterbarker Kind of. Previously we required that to arm you had absolute position available, now we only require relative position and that origin absolute position was set. But it's true that even with absolute position on arm we wouldn't guarantee that the polygon was loaded.

That's the least of the problems, the bug in the method checking if a point is inside the fence is much worse. Check my answer on the inline comment.

File: libraries/AC_Fence/AC_Fence.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AC_Fence/AC_Fence.cpp]: https://github.com/ArduPilot/ardupilot/pull/7410

**Comment by peterbarker** [comment]
@OXINARF Pre-arm checks for boundary validity: https://github.com/ArduPilot/ardupilot/pull/7411

**Pull request merged**